### PR TITLE
update to gptmd engine

### PR DESCRIPTION
**Pull request by trishorts**
effort to make gptmd more thread saferish

### CODE REVIEW

**Comment by trishorts** [comment]
performance of the master on top-down:
image

performance of the PR on top-down:
image

File: MetaMorpheus/EngineLayer/Gptmd/GptmdEngine.cs [hunk 1]
```diff
@@ -60,134 +62,124 @@ public static bool ModFits(Modification attemptToLocalize, IBioPolymer protein,
...
-                int localModsAdded = 0;
-
-                //foreach peptide in each psm and for each modification that matches the notch,
-                //add that modification to every allowed residue
-                //return those matches that give the highest score
-                for (; i < psms.Count(); i += maxThreadsPerFile)
+                return new GptmdResults(this, new Dictionary<string, HashSet<Tuple<int, Modification>>>(), 0);
+            }
+            else
+            {
```
Comment (nbollis) [inline MetaMorpheus/EngineLayer/Gptmd/GptmdEngine.cs:78]: Else appears to be redundant if above if returns

File: MetaMorpheus/EngineLayer/Gptmd/GptmdEngine.cs [hunk 2]
```diff
@@ -60,134 +62,124 @@ public static bool ModFits(Modification attemptToLocalize, IBioPolymer protein,
...
+                                            var scores = new List<double>();
+                                            var dissociationType = CommonParameters.DissociationType == DissociationType.Autodetect ?
+                                                psms[i].MsDataScan.DissociationType.Value : CommonParameters.DissociationType;
 
-                                        scores = CalculatePeptideScores(newPeptides, dissociationType, psms[i]);
+                                            scores = CalculatePeptideScores(newPeptides, dissociationType, psms[i]);
 
-                                        // If the score is within tolerance of the highest score, add the mod to the peptide
-                                        // If the tolerance is too tight, then the number of identifications in subsequent searches will be reduced
-                                        double scoreTolerance = 0.1;
```
Comment (nbollis) [inline MetaMorpheus/EngineLayer/Gptmd/GptmdEngine.cs:112]: Magic number should probably be a property

File: MetaMorpheus/EngineLayer/Gptmd/GptmdEngine.cs [hunk 3]
```diff
@@ -60,134 +62,124 @@ public static bool ModFits(Modification attemptToLocalize, IBioPolymer protein,
...
-
-            return new GptmdResults(this, modDict.ToDictionary(kvp => kvp.Key, kvp => new HashSet<Tuple<int, Modification>>(kvp.Value)), modsAdded);
+                });
+
+                // Convert ConcurrentDictionary to Dictionary with HashSet
+                var finalModDict = modDict.ToDictionary(
+                    kvp => kvp.Key,
+                    kvp => new HashSet<Tuple<int, Modification>>(kvp.Value)
+                );
+                return new GptmdResults(this, finalModDict, modsAdded);
```
Comment (nbollis) [inline MetaMorpheus/EngineLayer/Gptmd/GptmdEngine.cs:181]: Do you ever increment mods added?

**Review by nbollis — CHANGES_REQUESTED** [review]

File: MetaMorpheus/EngineLayer/Gptmd/GptmdEngine.cs (hunk 3 above)
Comment (trishorts) [reply MetaMorpheus/EngineLayer/Gptmd/GptmdEngine.cs:185]: lol. no. but it is not incremented in the master. i just fixed this.

File: MetaMorpheus/EngineLayer/Gptmd/GptmdEngine.cs (hunk 2 above)
Comment (trishorts) [reply MetaMorpheus/EngineLayer/Gptmd/GptmdEngine.cs:112]: done

**Review by nbollis — APPROVED** [review]

**Review by Alexander-Sol — APPROVED** [review]

**Pull request merged**